### PR TITLE
Updates to support u-muP, as the new default behaviour

### PR DESCRIPTION
**Pull request by DouglasOrr**
[Demo notebook](https://github.com/graphcore-research/unit-scaling/blob/umup-updates/examples/demo.ipynb).

[README header](https://github.com/graphcore-research/unit-scaling/tree/umup-updates?tab=readme-ov-file#unit-scaled-maximal-update-parameterization-u-%CE%BCp).

Changes:
 - Add mults to nonlinear `U.*` ops and `uu.*` modules
 - Add `U.readout_linear`, `uu.ReadoutLinear` to apply 1/fan_in scaling
 - Add `U.rms_norm`, `uu.RMSNorm`, `U.silu`, `uu.SiLU`, `U.silu_glu`, `U.mse_loss`
 - Add `uu.Parameter`, `uu.optim.*` to provide LR scaling
 - Add `uu.Trunk`, `uu.TransformerStack` to set `mup_scaling_depth` and (for `TransformerStack`) apply appropriate residual mults
 - Add demo notebook `examples/demo.ipynb`
 - Change the default constraint from "gmean" -> "to_output_scale"
 - Change default `uu.Linear` to `bias=False` & `uu.*Norm` to `elementwise_affine=False`
 - Change `U.softmax` and `U.scaled_dot_product_attention` scaling rules to use the empirical fit
 - Change definition of `tau` in `residual_split`/`residual_add`
 - Change `uu.MHSA` to use `U.scaled_dot_product_attention`
 - Change `uu.MLP` to use SwiGLU
 - Change `uu.TransformerLayer`, `uu.TransformerDecoder` to use `RMSNorm` and various other tweaks
 - Update top of README (other docs updates are lagging)
 - +probably more

### CODE REVIEW

**Comment by DouglasOrr** [comment]
@lyprince, thank you for agreeing to review!

@thecharlieblake FYI

This is still WIP, but (I think/hope) the major changes are in.

File: examples/scale_analysis.py [hunk 1]
```diff
@@ -49,7 +51,7 @@
 backward = torch.randn(batch_size, seq_len, hidden_size)
 
 annotated_code = analyse_module(
-    uu.TransformerLayer(hidden_size, heads, dropout_p), input, backward
+    uu.TransformerLayer(hidden_size, heads, 0.1, 1.0, False, dropout_p), input, backward
```
Comment (lyprince) [inline examples/scale_analysis.py:54]: Nit: could you replace `0.1, 1.0, False` with variable names?

File: unit_scaling/_modules.py [hunk 2]
```diff
@@ -4,43 +4,75 @@
...
     variadic_constraint_docstring,
 )
+from .parameter import MupType, Parameter, has_parameter_data
 
 
 @inherit_docstring(
     short_description="Applies a **unit-scaled** Gaussian Error Linear Units function:",
-    add_args=[binary_constraint_docstring],
+    add_args=[mult_docstring(), binary_constraint_docstring],
     unsupported_args=["approximate"],
```
Comment (lyprince) [inline unit_scaling/_modules.py:29]: Not critical, but is there any reason this can't support approximate? Surely the approximation is good enough that the `mult` can be the same whether using `erf` or `tanh` implementation.

File: unit_scaling/_modules.py [hunk 3]
```diff
@@ -91,13 +128,17 @@ def __init__(
         self,
         in_features: int,
         out_features: int,
-        bias: bool = True,
+        bias: bool = False,
```
Comment (lyprince) [inline unit_scaling/_modules.py:131]: why do you deviate from base pytorch here?

File: unit_scaling/_modules.py [hunk 4]
```diff
@@ -198,35 +399,36 @@ class MHSA(nn.Module):
...
         self.dropout_p = dropout_p
-        self.linear_qkv = Linear(
-            hidden_size, 3 * hidden_size, bias=False, constraint=constraint
-        )
-        self.linear_o = Linear(
-            hidden_size, hidden_size, bias=False, constraint=constraint
-        )
+        self.is_causal = is_causal
+        self.mult = mult
+        self.linear_qkv = Linear(hidden_size, 3 * hidden_size, constraint=constraint)
```
Comment (lyprince) [inline unit_scaling/_modules.py:422]: Without a `linear` specialisation (like `linear_readout`), this will have an extra `sqrt(3)` factor in the scale.

File: unit_scaling/_modules.py [hunk 5]
```diff
@@ -241,54 +443,111 @@ class TransformerLayer(nn.Module):
...
         input = self.mlp(input)
         input = U.dropout(input, self.dropout_p, self.training)
         return U.residual_add(input, skip, tau=self.mlp_tau)
 
 
+@inherit_docstring(
+    "A :class:`torch.nn.Sequential` that automatically configures the depth"
+    " for sake of scaling."
+)
+class Trunk(nn.Sequential):
```
Comment (lyprince) [inline unit_scaling/_modules.py:502]: using `nn.Sequential` is fine when only passing activations between blocks in the trunk, but typically we pass a lot more (`masks`, `embeddings`). 

Worth considering using a more general `ModuleList` implementation, perhaps with some requirement that elements in list are unit scaled blocks.

File: unit_scaling/_modules.py [hunk 6]
```diff
@@ -241,54 +443,111 @@ class TransformerLayer(nn.Module):
...
+                    mlp_tau=residual_scaling(2 * i + 1, 2 * layers),
+                )
+                for i in range(layers)
+            )
+        )
+
+
 @format_docstring(variadic_constraint_docstring)
-class TransformerDecoder(nn.Module):  # pragma: no cover
+class TransformerDecoder(nn.Sequential):  # pragma: no cover
```
Comment (lyprince) [inline unit_scaling/_modules.py:546]: Same problems using `nn.Sequential` here. We'd typically register tensor data as buffers in the root module that we want to reuse across layers, which is difficult with `nn.Sequential`

File: unit_scaling/_modules.py [hunk 7]
```diff
@@ -313,34 +573,34 @@ def __init__(
...
-        self.dropout_p = dropout_p
-        self.initial_layer_norm = LayerNorm(hidden_size)
-        self.transformer_layers = nn.Sequential(
-            *(
-                TransformerLayer(
-                    hidden_size, heads, dropout_p, act_fn, mhsa_tau, mlp_tau, constraint
-                )
-                for _ in range(layers)
-            )
+        self.embedding_dropout = Dropout(dropout_p)
```
Comment (lyprince) [inline unit_scaling/_modules.py:584]: I don't think we have dropout or norm after embedding usually? (based on https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama/model.py#L417-L436)

File: unit_scaling/functional.py [hunk 8]
```diff
@@ -247,52 +365,61 @@ def residual_split(input: Tensor, tau: float = 0.5) -> Tuple[Tensor, Tensor]:
...
-    standard transformer. Empirically, the self-attention tau is fairly insensitive
-    (i.e. tau=0.1 or tau=0.001 work well), but the default tau=0.5 causes significant
-    degradation.
-
     Args:
         input (Tensor): the tensor to which the residual layer is to be applied.
-        tau (float, optional): the weighting of the residual branch relative to the skip
-            connection. Defaults to 0.5.
+        tau (float, optional): the ratio of scale of contributions of the residual
+            branch to the skip connection. Larger values favor skip over residual.
```
Comment (lyprince) [inline unit_scaling/functional.py:371]: Should this be "values larger than one favor skip over residual"?

File: unit_scaling/functional.py [hunk 9]
```diff
@@ -247,52 +365,61 @@ def residual_split(input: Tensor, tau: float = 0.5) -> Tuple[Tensor, Tensor]:
...
+            Defaults to 1 (equal contribution).
 
     Returns:
         Tuple[Tensor, Tensor]: resulting tensors in the order: `residual, skip`.
     """
-    residual = scale_bwd(input, tau**0.5)
-    skip = scale_bwd(input, (1 - tau) ** 0.5)
+    denom = (1 + tau**2) ** 0.5
```
Comment (lyprince) [inline unit_scaling/functional.py:377]: Do you need to explain the `denom` in the docstring? Or is this just a new way to scale so that the default can be 1.0?

Comment (lyprince) [reply unit_scaling/functional.py:377]: Reading through the paper I think I understand this better now. It might be useful to explicitly reference the section in the paper justifying this parameterization (Appendix G I think?)

File: unit_scaling/optim.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+                    params=[param],
+                    lr=param_lr,
+                    weight_decay=param_weight_decay,
+                    **{
+                        k: v
+                        for k, v in group.items()
+                        if k not in ("params", "lr", "weight_decay")
+                    },
+                )
+            )
```
Comment (lyprince) [inline unit_scaling/optim.py:143]: Non-essential here, but worth considering a pass that groups together parameters with the same lr and weight decay, as this allows `foreach` functions used in optimizers to take effect

File: unit_scaling/optim.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    return result
+
+
+@inherit_docstring(
+    short_description="An **lr-scaled** version of :class:`torch.optim.SGD` for u-muP."
+)
+class SGD(torch.optim.SGD):
+    def __init__(
+        self,
+        params: ParamsT,
```
Comment (lyprince) [inline unit_scaling/optim.py:153]: I'm guessing you can't pass a dictionary of pre-grouped parameters (you can for base torch optimizers)?

File: unit_scaling/functional.py [hunk 12]
```diff
@@ -126,7 +192,7 @@ def dropout(
 def matmul(
     left: Tensor,
     right: Tensor,
-    constraint: Optional[str] = "gmean",
+    constraint: Optional[str] = "to_output_scale",
```
Comment (lyprince) [inline unit_scaling/functional.py:195]: Why the change to `to_output_scale` as default for matmul too?

File: unit_scaling/optim.py (hunk 11 above)
Comment (lyprince) [reply unit_scaling/optim.py:153]: Ignore this - I see a test further down

**Review by lyprince — COMMENTED** [review]
Looks good! I've done my best to check for correctness. I have some uncertainty around residual scaling. A test for unit scale preserved across depth with the residual scaling scheme would give me more confidence.

Other than that, my main comments are around 
1) Documentation that points to relevant parts of paper for justifying design choices / changes.
2) Doubts over whether use of `nn.Sequential` is the right choice for keeping track of depth, given common patterns for implementing mask and positional embeddings.
3) Clarification on whether it is intentional to fuse qkv projection in MHSA.
4) Testing compatibility of Parameter class with `torch.compile` (does FakeTensor trigger `deepcopy` failure).

File: unit_scaling/core/functional.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    mean = x.float().pow(2).mean(dims, keepdim=keepdim)
+    if eps:
+        mean = mean + eps
+    return mean.sqrt().to(x.dtype)
+
+
+ResidualScalingFn = Callable[[int, int], float]
+
+
+def transformer_residual_scaling_rule(
```
Comment (lyprince) [inline unit_scaling/core/functional.py:83]: Can you test that this preserves unit scale in the Transformer model?

File: unit_scaling/parameter.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+def has_parameter_data(parameter: nn.Parameter) -> TypeGuard[ParameterData]:
+    """Check that the parameter supports the :class:`ParameterData` protocol."""
+    return (
+        getattr(parameter, "mup_type", None) in MupType.__args__
+        and hasattr(parameter, "mup_scaling_depth")
+        and isinstance(parameter.mup_scaling_depth, (type(None), int))
+    )
+
+
+def _parameter_deepcopy(self: nn.Parameter, memo: Dict[int, Any]) -> nn.Parameter:
```
Comment (lyprince) [inline unit_scaling/parameter.py:37]: Can you test whether this `deepcopy` override is compatible with FakeTensor use in torch.compile?

File: unit_scaling/_modules.py (hunk 7 above)
Comment (DouglasOrr) [reply unit_scaling/_modules.py:584]: Agree, thanks - a shared-embedding hangover I think, I've deleted it.

File: unit_scaling/_modules.py (hunk 2 above)
Comment (DouglasOrr) [reply unit_scaling/_modules.py:29]: I agree, should be close enough; I'll remove this check.

File: unit_scaling/_modules.py (hunk 4 above)
Comment (DouglasOrr) [reply unit_scaling/_modules.py:422]: Ah, yes, I wondered about this; the cheeky thing I was thinking is that when using the new-default constraint "to_output_scale" it shouldn't matter, as the scale just depends on fan_in. But maybe this is a bit cheeky as it can be overridden.

File: unit_scaling/_modules.py (hunk 5 above)
Comment (DouglasOrr) [reply unit_scaling/_modules.py:502]: Agree that this is needed; I've added it alongside, since I think both are valid. Not entirely happy with the name `DepthModuleList`. I wanted to convey that not every `ModuleList` should use this impl, only ones that have a depth-hyperparameter-scaling behaviour, but it seems a bit awkward.

> Perhaps with some requirement that elements in list are unit scaled blocks.

This should already be enforced by the `has_parameter_data` check, I think.

File: unit_scaling/_modules.py (hunk 6 above)
Comment (DouglasOrr) [reply unit_scaling/_modules.py:546]: Agree; we can switch to regular if/when we do this. TBH I'm not sure about such high-level modules "just for demonstrating scaling" continuing to live inside `uu.*`, but thought it worth keeping them for now, at least for sake of illustration.

File: unit_scaling/functional.py (hunk 9 above)
Comment (DouglasOrr) [reply unit_scaling/functional.py:377]: :+1:, might leave it until the numbering is definitively frozen (very soon!), but yes.

File: unit_scaling/optim.py (hunk 10 above)
Comment (DouglasOrr) [reply unit_scaling/optim.py:143]: :+1: agree!

File: unit_scaling/functional.py (hunk 12 above)
Comment (DouglasOrr) [reply unit_scaling/functional.py:195]: Good point.

I suppose if a client used `matmul` to implement something like `scaled_dot_product_attention` manually, they'd be more likely to get the right behaviour (consistent with the new unit-scaling principle) using `to_output_scale` than `gmean`. Not sure if that outweighs the "matmul is a general tool" thing, though.

File: unit_scaling/core/functional.py (hunk 13 above)
Comment (DouglasOrr) [reply unit_scaling/core/functional.py:83]: I think `residual_split`/`residual_add` preserve unit scale whatever the `tau`, so the test ends up being whether it keeps the promise of equal-contribution for each layer. I think the maths of the `test_transformer_residual_scaling_rule` does this job (even though it just uses scalars), but definitely open to more testing if I'm missing something.

File: unit_scaling/parameter.py (hunk 14 above)
Comment (DouglasOrr) [reply unit_scaling/parameter.py:37]: Thanks, yes, I've tried to write a test using `torch.compile`, but I'm not sure if it exercises the FakeTensor path you mention.

File: unit_scaling/_modules.py (hunk 3 above)
Comment (DouglasOrr) [reply unit_scaling/_modules.py:130]: As discussed, I guess the deviation seems the lesser of two evils.

File: examples/demo.ipynb [hunk 15]
```diff
@@ -0,0 +1,625 @@
...
+    "from torch import nn, Tensor, tensor\n",
+    "import torch.nn.functional as F\n",
+    "\n",
+    "import unit_scaling as uu\n",
+    "import unit_scaling.functional as U\n",
+    "\n",
+    "# Config & helpers\n",
+    "torch.backends.cuda.matmul.allow_tf32 = True\n",
+    "torch.backends.cudnn.allow_tf32 = True\n",
+    "def show_layer_stats(layer: nn.Module, input_shape: Tuple[int, ...]) -> None:\n",
```
Comment (lyprince) [inline examples/demo.ipynb:52]: Demo notebook looks good! Only thing that broke my attention was searching for `show_layer_stats`. Was  easy to miss when colocated with imports. My preference would be to have it defined next to first use, or at least be in its own cell.

File: unit_scaling/_modules.py (hunk 4 above)
Comment (lyprince) [reply unit_scaling/_modules.py:418]: > the scale just depends on fan_in
Won't the output scale of this op be `3 * fan_in` rather than fan_in though? I see in the demo notebook that `attn_qkv.weight.grad.std = 0.62`, which is pretty close to `1/sqrt(3) = 0.58`.
